core/src/apps/cardano/sign_tx.py

```diff
@@ -117,12 +139,41 @@ def _validate_outputs(
         raise wire.ProcessError("Total transaction amount is out of range!")
 
 
+def _validate_certificates(certificates: List[CardanoTxCertificateType]) -> None:
+    for certificate in certificates:
+        if not is_staking_path(certificate.path):
+            raise INVALID_CERTIFICATE
+
+        if certificate.type == CardanoCertificateType.STAKE_DELEGATION:
+            if certificate.pool is None or len(certificate.pool) != POOL_HASH_SIZE:
+                raise INVALID_CERTIFICATE
+
+
+def _validate_withdrawals(withdrawals: List[CardanoTxWithdrawalType]) -> None:
+    for withdrawal in withdrawals:
+        if not is_staking_path(withdrawal.path):
+            raise INVALID_WITHDRAWAL
+
+        if not 0 <= withdrawal.amount < LOVELACE_MAX_SUPPLY:
+            raise INVALID_WITHDRAWAL
+
+
 def _serialize_tx(keychain: seed.Keychain, msg: CardanoSignTx) -> Tuple[bytes, bytes]:
     tx_body = _build_tx_body(keychain, msg)
     tx_hash = _hash_tx_body(tx_body)
 
-    witnesses = _build_witnesses(keychain, msg.inputs, tx_hash, msg.protocol_magic)
+    witnesses = _build_witnesses(
```

Comment:
I see no reason for having so many arguments to _build_witnesses. I think I would separate different witness creation logic

Reply:
I like the fact that the witness creation is in one separate place at the moment. I can reduce the number of arguments by passing the whole message in instead. ... What do you mean by different witness creation? Byron vs. Shelley? Or Byron vs. Input witnesses vs. Certificate witnesses vs. Withdrawal witnesses? I don't like the fact that then the caller has to assemble the witnesses. I think this is again worth taking offline.

Reply:
Discussed offline - we'll keep it as is, since the caller would have to do extra work to build the witnesses.